src/main/java/dev/morling/onebrc/CalculateAverage_hundredwatt.java

```diff
@@ -0,0 +1,339 @@
+        // fix two ';' in word issue here (rather than in hot path)
+        byte b;
+        int limit = kb.position() - 8;
+        kb.position(limit);
+        while ((b = kb.get()) != 0 && b != ';') {
+            limit++;
+        }
+
+        kb.flip();
+        byte[] bytes = new byte[limit];
+        kb.get(bytes);
+
+        return new String(bytes).replace("\0", "");
+    }
+
+    private static Record merge(Record v, Record value) {
+        var record = new Record();
+        record.min = (short) Math.min(v.min, value.min);
+        record.max = (short) Math.max(v.max, value.max);
+        record.sum = v.sum + value.sum;
+        record.count = v.count + value.count;
+        return record;
+    }
+
+    private static int processChunk(ByteBuffer bb, HashTable hashTable, long start, long size) {
+        // Find first entry
+        while (start != 0 && bb.get() != '\n') {
+        }
+
+        long word;
+        long[] key = new long[13];
+        int offset;
+        long arg, hasvalue, op1, op2;
+        int position = bb.position();
+        long hash;
+        long temperature_hash;
+        int temperature_pos;
+        short temperature_value;
+
+        int i = 0;
+        int end = (int) (size - MAX_ROW_SIZE);
+        while (position < end) {
+            i++;
+            offset = -1;
+
+            // Parse city name
+            // First word
+            hash = key[++offset] = bb.getLong(position + offset * 8);
+
+            // From "Determine if a word has a byte equal to n"
+            // https://graphics.stanford.edu/~seander/bithacks.html#ValueInWord
+            arg = (key[offset]) ^ (0x0101010101010101L * (';'));
+            op1 = (arg - 0x0101010101010101L);
+            op2 = ~(arg);
+            hasvalue = (op1 & op2 & 0x8080808080808080L);
+
+            // Remaining words (if present)
+            while (hasvalue == 0) {
+                ++offset;
+                key[offset] = bb.getLong(position + offset * 8);
+                hash ^= key[offset];
+
+                arg = (key[offset]) ^ (0x0101010101010101L * (';'));
+                op1 = (arg - 0x0101010101010101L);
+                op2 = ~(arg);
+                hasvalue = (op1 & op2 & 0x8080808080808080L);
+            }
+            hash ^= key[offset]; // unset last word since it will be updated
+            key[offset] = key[offset] & (-hasvalue); // ';' == 0x3b and -hasvalue is something like 0xff8000..., we can ignore the 0x80 byte since 0x3b & 0x80 == 0
+            hash ^= key[offset];
+
+            position = position + offset * 8 + (Long.numberOfLeadingZeros(hasvalue)) / 8 + 1; // +1 for \n
+
+            // Parse temperature
+            word = bb.getLong(position);
+            arg = (word) ^ 0x0101010101010101L * ('\n');
+            op1 = (arg - 0x0101010101010101L);
+            op2 = ~(arg);
+            hasvalue = (op1 & op2 & 0x8080808080808080L);
+            word = word & ((-hasvalue));
```

Comment:
If you load the value in little-endian, and since all characters before the newline are positives larger than `0x0A` (newline), I believe you can just do

   (word - 0x0B0B0B0B0B0B0B0BL) & 0x8080808080808080A

The lowest set bit of this will belong to the byte `\n`

Reply:
Thanks for the suggestion!

I was able to verify it works in a single test case 😄

```java
    byte[] bytes = new byte[] { 0x2d, 0x36, 0x37, 0x2e, 0x38, 0x0a, 0x78, 0x78 };
    ByteBuffer bb = ByteBuffer.wrap(bytes).order(ByteOrder.LITTLE_ENDIAN);
    bb.order();
    long word = bb.getLong(0);
    hasvalue = (word - 0x0B0B0B0B0B0B0B0BL) & 0x8080808080808080L;
    System.out.printf("%16s\n", Long.toHexString(word));
    System.out.printf("%16s\n", Long.toHexString(hasvalue));
    System.out.printf("%16s\n", Long.toHexString(-hasvalue));
    System.out.printf("%16s\n", Long.toHexString(word & (~(-hasvalue))));
```

```
System.out.printf("%16s\n", Long.toHexString(word)) = java.io.PrintStream@5a10411
78780a382e37362d


System.out.printf("%16s\n", Long.toHexString(hasvalue)) = java.io.PrintStream@5a10411
    800000000000


System.out.printf("%16s\n", Long.toHexString(-hasvalue)) = java.io.PrintStream@5a10411
ffff800000000000


System.out.printf("%16s\n", Long.toHexString(word & (~(-hasvalue)))) = java.io.PrintStream@5a10411
     a382e37362d
```

Reply:
Little-endian matches the processor too...

Excited to try refactoring my fork to use this, thanks again! 😄

Reply:
@merykitty My first attempt at your approach yields a slight improvement! I may be able to simplify further too:

h2 - is the fork with your suggestion
hundredwatt - is the current implementation

<img width="698" alt="image" src="https://github.com/gunnarmorling/1brc/assets/91577/dfeb62ac-f4b0-4b18-9264-069389b185ed">

1.5% faster, but noisy

Reply:
https://github.com/gunnarmorling/1brc/pull/338/files#diff-66aeaf68fbf201bdaeae12980f29e9baa53465c8410f0988869c7fd8356b14f8L268